### ISSUE TITLE
Proposal: Better Shopping Cart Markup.

### ISSUE THREAD

**Issue by georgestephanis** [woothemes/woocommerce#8708]
tl;dr: Stop generating the Shopping Cart (widget and page) with PHP, and just load in cart data to generate it with JS.

Long version:

This will have a couple wins:

_Better compatibility with full-page caching._

(this win is specifically for logged out users)

By generating the cart based on data in sessionStorage (or localStorage, but sessionStorage feels better for this) most pages can be stored by a full page caching plugin, such as SuperCache, BatCache, W3 Total Cache, etc.  Then the cart data (for logged out users) can be pulled either from sessionStorage directly, or via ajax and then cached in sessionStorage.

_More Accurate Cart Info._

We've all been there.  When browsing a store, we open multiple products in different tabs, add something to a cart in one tab, then switch back to another tab only to find that the cart is out of date.  By using JS to generate our Shopping Cart widget and Cart pages, we can listen for changes in other tabs, and then update all open tabs markup to reflect the new cart contents.  Implementation details: https://truongtx.me/2014/06/16/cross-tab-communication-using-html5-dom-storage/

---

If all we care about is the second win, we could leave the PHP cart generation in place, but that would be something of a duplication of effort as going forward we'd need to keep the PHP generation and JS template generation in sync, but by leaving the cart markup in the php, it could get cached by a full page cache provider.  I'd like to see it removed entirely in lieu of purely JS generated carts.

**Comment by georgestephanis** [comment]
Correction: It looks like the shopping cart contents is already being built with Javascript, not PHP.  This should be much easier than I expected.

https://github.com/woothemes/woocommerce/blob/master/includes/widgets/class-wc-widget-cart.php#L65-L66

**Comment by mikejolley** [comment]
@georgestephanis The cart widget is refreshed (on page load) when the cart contents changes (we use a hash to detect this). Session storage is used to prevent an ajax load each time. This busts caching.

The markup of the widget is in PHP mainly to allow customisation via themes and hooks/filters.

**Comment by georgestephanis** [comment]
@mikejolley Yup, I see that now.  We could still add something to update across multiple tabs when the sessionStorage is updated with the new php-generated markup, though.

**Comment by mikejolley** [comment]
That would be a decent enhancement. Want to give it a shot in a branch for 2.5?

**Comment by georgestephanis** [comment]
Unsure if I'll have time, but I'd be happy to give it a whack.

Alternately, cc: @nb @mdawaffe if it looks like it may be a good trial project maybe?  Pretty self encapsulated, and will require some js chops.  Not sure if it's sizable enough to be a trial project though.

**Comment by nb** [comment]
It’s fine if projects are a bit smaller. Could you tag potential trial projects with a “Trial Project” label?

**Comment by georgestephanis** [comment]
@nb Sorry for the delay, label created and added.  @mikejolley said they have a couple other possibilities in mind on  the Trello board, but want to make sure having it as a potential trial project doesn't mean that other folks are hesitant to actually clobber it in the meanwhile.

**Comment by nb** [comment]
I am find with any other means of marking it, hit me privately if you think the label is not the best option.
